### DISCUSSION TITLE
GitHub actions ccache and cache misses

### DISCUSSION THREAD

**Discussion by silverqx** [ccache/ccache#1241] [Support]
Hi, I have big cache misses if I compile on GitHub actions, practically all platforms and compilers, Linux/Windows, and GCC/Clang. Misses are like 50-80%.

I can achieve 100% cache hits on my development machine, but not on GitHub actions CI.

My steps look like this eg. for Linux:
```yml
    - name: Ccache initialize
      id: initialize-ccache-cache
      run: |
        cachePath=$(ccache --get-config cache_dir)
        echo "CachePath=$cachePath" >> $GITHUB_OUTPUT

        echo "ImageOS=$ImageOS" >> $GITHUB_OUTPUT

    - name: Ccache restore cache 🕺
      uses: actions/cache@v3
      with:
        path: ${{ env.cache_path }}
        key: ${{ runner.os }}-${{ env.image_os }}-ccache-${{ env.cache_name }}-${{ github.run_id }}
        restore-keys: |
          ${{ runner.os }}-${{ env.image_os }}-ccache-${{ env.cache_name }}-
      env:
        cache_name: ${{ matrix.compiler.key }}-qt62
        cache_path: ${{ steps.initialize-ccache-cache.outputs.CachePath }}
        image_os: ${{ steps.initialize-ccache-cache.outputs.ImageOS }}

    - name: Ccache setup 🥳
      run: |
        echo '::group::Prepare ccache config'
        ccache --set-config compression=false
        ccache --set-config max_size=${{ matrix.compiler.key == 'gcc12' && '400M' || '250M' }}
        ccache --set-config sloppiness=pch_defines,time_macros
        ccache --show-config
        echo '::endgroup::'

        echo '::group::Clear ccache statistics'
        ccache --zero-stats
        echo '::endgroup::'

        echo '::group::Print version'
        ccache --version
        echo '::endgroup::'
```

Now I'm trying `sloppiness=include_file_ctime,include_file_mtime,pch_defines,time_macros` if it doesn't help.

I'm posting a question here because I don't understand where the problem can be or what is causing this cache misses. 

I think that the problem will be because all source code, dependencies, and also the ccache folder (cache_dir) are re-created for every run and they have different ctime and mtime?

What do you think? Where the problem can be and how can it be workarounded?

**Comment by jrosdahl** [comment]
It's hard to tell without more details, but here are some guesses:

* Your build system uses absolute paths to source files (different for different GitHub runners?) and those end up in the input hash.
* Your build system uses `-g` or similar and ccache therefore includes the current working directory (different for different GitHub runners?) in the input hash.

See [_Compiling in different directories_](https://ccache.dev/manual/4.7.4.html#_compiling_in_different_directories) for hints on how to work around the above issues. Also, read [_Cache debugging_](https://ccache.dev/manual/4.7.4.html#_cache_debugging) to learn how to check what the actual problem is.

**Comment by silverqx** [comment]
@jrosdahl Thx for reply, I have checked it and on GitHub actions are all paths still the same, they never change even for all dependencies, eg. vcpkg or standard folders on Linux.

I spent all of yesterday on it, I have setup debugging and find out that the first build generated some manifest or preprocessed hashes and the second build could not find these hashes, the hashes were absolutely the same, first I was confused, but then I realized that it means that my `max_size` is low because I have disabled `compression`.

So I did a clean build for every supported compiler and checked the result with `ccache --show-compression` to find out how much cache ccache needs.

**Reply by jrosdahl** [reply]
>  I have disabled `compression`.

Why disable compression?

> So I did a clean build for every supported compiler and checked the result with ccache --show-compression to find out how much cache ccache needs.

You can use `ccache -s` to see the current cache size right away: if _Cleanups_ is `0` and you started with an empty cache, the _Cache size (GB)_ number tells you right away what's needed. You don't need to run `ccache --show-compression` to tell how space is needed unless you plan on converting your cache to not use compression.

**Comment by silverqx** [comment]
> Why disable compression?

Maybe I drank too much coffee 🙃, I wanted to avoid redundant compression, but I was totally wrong about it 😵‍💫, but now everything works great and I have 100% direct hits on all platforms, so happy end 😁, thank you for replies
